test/ProgramScopeConstants/issue_102.cl

```diff
@@ -0,0 +1,71 @@
+// CHECK:  [[_v4float:%[0-9a-zA-Z_]+]] = OpTypeVector [[_float]] 4
+// CHECK:  [[__struct_20:%[0-9a-zA-Z_]+]] = OpTypeStruct [[_v4float]] [[_float]] {{.*}}
+// CHECK:  [[_uint_2:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 2
+// CHECK:  [[__arr__struct_20_uint_2:%[0-9a-zA-Z_]+]] = OpTypeArray [[__struct_20]] [[_uint_2]]
+// CHECK:  [[%_ptr_Private__arr__struct_20_uint_2:%[0-9a-zA-Z_]+]] = OpTypePointer Private [[__arr__struct_20_uint_2]]
+// CHECK:  [[__struct_28:%[0-9a-zA-Z_]+]] = OpTypeStruct [[_v4float]] [[_float]]
+// CHECK:  [[_uint_2_0:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 2
+// CHECK:  [[__arr__struct_28_uint_2_0:%[0-9a-zA-Z_]+]] = OpTypeArray [[__struct_28]] [[_uint_2_0]]
+// CHECK:  [[%_ptr_Private__arr__struct_28_uint_2_0:%[0-9a-zA-Z_]+]] = OpTypePointer Private [[__arr__struct_28_uint_2_0]]
+// CHECK:  [[_uint_0:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 0
+// CHECK:  [[_float_10_5:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 10.5
+// CHECK:  [[_float_11_5:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 11.5
+// CHECK:  [[_float_12_5:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 12.5
+// CHECK:  [[_float_13_5:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 13.5
+// CHECK:  [[_x:_%[0-9a-zA-Z_]+]] = OpConstantComposite [[_v4float]] [[_float_10_5]] [[_float_11_5]] [[_float_12_5]] [[_float_13_5]]
+// CHECK:  [[_float_14_5:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 14.5
+// CHECK:  [[_y:%[0-9a-zA-Z_]+]] = OpConstantComposite [[__struct_28]] [[_x]] [[_float_14_5]]
+// CHECK:  [[_float_20_5:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 20.5
+// CHECK:  [[_float_21_5:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 21.5
+// CHECK:  [[_float_22_5:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 22.5
+// CHECK:  [[_float_23_5:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 23.5
+// CHECK:  [[_z:%[0-9a-zA-Z_]+]] = OpConstantComposite [[_v4float]] [[_float_20_5]] [[_float_21_5]] [[_float_22_5]] [[_float_23_5]]
+// CHECK:  [[_float_24_5:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 24.5
+// CHECK:  [[_w:%[0-9a-zA-Z_]+]] = OpConstantComposite [[__struct_28]] [[_z]] [[_float_24_5]]
+// CHECK:  [[_t%[0-9a-zA-Z_]+]] = OpConstantComposite [[__arr__struct_28_uint_2_0]] [[_y]] [[_w]]
+// CHECK:  [[_uint_1:%[0-9a-zA-Z_]+]] = OpConstant [[_uint]] 1
+// CHECK:  [[_float_1:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 1
+// CHECK:  [[_float_2:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 2
+// CHECK:  [[_float_3:%[0-9a-zA-Z_]+]] = OpConstant [[_float]] 3
+// CHECK:  [[_u:%[0-9a-zA-Z_]+]] = OpConstantComposite [[__arr_float_uint_3]] [[_float_1]] [[_float_2]] [[_float_3]]
+// CHECK:  [[_pu:%[0-9a-zA-Z_]+]] = OpVariable [[_ptr_Private__arr_float_uint_3]] Private [[_u]]
+// CHECK:  [[_pt:%[0-9a-zA-Z_]+]] = OpVariable [[_ptr_Private__arr__struct_28_uint_2_0]] Private [[_t]]
+// CHECK:  [[_pf:%[0-9a-zA-Z_]+]] = OpAccessChain {{.*}} [[_pu]]
+// CHECK:  [[_f:%[0-9a-zA-Z_]+]] = OpLoad [[_float]] [[_pf]]
+// CHECK:  [[_bc:%[0-9a-zA-Z_]+]] = OpBitcast [[_ptr_private__arr__struct_20_uint_2]] [[_pt]]
```

Comment:
Ok I guess this wasn't immediately obvious to me previously, but this is invalid SPIR-V. Shaders don't support this type of bitcast. It is unfortunate that the validator is not catching this, but I think https://github.com/KhronosGroup/SPIRV-Tools/pull/3091 should flag this as invalid.

This is a case that should be handled by ReplacePointerBitcastsPass. Instead of casting the variable, clspv would have to load 4 uints, make a vector and bitcast that vector to a float vector.

Reply:
Bummer. For what it's worth, the SPIR-V spec reads a bit confusingly (IMHO) here. Looking at OpBitcast from here: https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.pdf does not convey any such limitation. Furthermore, there are only 5 references to OpBitcast in the entire document, neither of which makes reference to any such limitation being in place. Yes, 2.2.2 is normative in what constitutes a valid source of a variable pointer, but I'd have hoped it'd be reflected under OpBitcast too.  

Granted, it appears to be normative in 2.16.1, and that outlines a more profound problem in that "A variable pointer must point to one of the following storage classes: StorageBuffer, Workgroup [...]", which is broken orthogonally to the bitcast, I think. So, TL;DR, hoisting the bitcast replacement / insertion before running ReplacePointerBitcastsPass wouldn't be a problem, but this latter aspect appears a bit more involved (to me), since it hints that mapping `__constant` to `Private` might not be a generic solution. Should I handle that separately?